exonum/src/storage/sparse_list_index.rs

```diff
@@ -0,0 +1,838 @@
+// Copyright 2017 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of array list of items with spaces.
+
+use byteorder::{BigEndian, ByteOrder};
+
+use std::borrow::Cow;
+use std::cell::Cell;
+use std::marker::PhantomData;
+
+use crypto::{hash, Hash};
+use super::{BaseIndex, BaseIndexIter, Snapshot, Fork, StorageValue};
+
+#[derive(Debug, Default, Clone, Copy)]
+struct SparseListSize {
+    /// Total list's length including spaces. In fact points to the next index for a new element.
+    capacity: u64,
+    /// Amount of non-empty elements.
+    length: u64,
+}
+
+impl SparseListSize {
+    fn to_array(&self) -> [u8; 16] {
+        let mut buf = [0; 16];
+        BigEndian::write_u64(&mut buf[0..8], self.capacity);
+        BigEndian::write_u64(&mut buf[8..16], self.length);
+        buf
+    }
+}
+
+impl StorageValue for SparseListSize {
+    fn hash(&self) -> Hash {
+        hash(&self.to_array())
+    }
+
+    fn into_bytes(self) -> Vec<u8> {
+        self.to_array().to_vec()
```

Comment:
...So you do not need `to_array` method and you can write bytes directly into a `Vec` here.